core/src/main/scala/org/ensime/indexer/domain.scala

```diff
@@ -195,61 +198,95 @@ final case class Descriptor(params: List[DescriptorType], ret: DescriptorType) {
     "(" + params.map(_.internalString).mkString("") + ")" + ret.internalString
 }
 
+sealed trait RawSymbol {
+  def fqn: String
+  def internalRefs: Set[FullyQualifiedName]
+}
+
 final case class RawClassfile(
-  name: ClassName,
-  generics: Option[GenericClass],
-  superClass: Option[ClassName],
-  interfaces: List[ClassName],
-  access: Access,
-  deprecated: Boolean,
-  fields: Queue[RawField],
-  methods: Queue[RawMethod],
-  source: RawSource
-)
+    name: ClassName,
+    generics: Option[GenericClass],
+    innerClasses: Set[ClassName],
+    superClass: Option[ClassName],
+    interfaces: List[ClassName],
+    access: Access,
+    deprecated: Boolean,
+    fields: List[RawField],
+    methods: Queue[RawMethod],
+    source: RawSource,
+    isScala: Boolean,
+    internalRefs: Set[FullyQualifiedName]
+) extends RawSymbol {
+  override def fqn: String = name.fqnString
+}
 
 final case class RawSource(
   filename: Option[String],
   line: Option[Int]
 )
 
-final case class RawType(
-  fqn: String,
-  access: Access
-)
-
 final case class RawField(
-  name: FieldName,
-  clazz: DescriptorType,
-  generics: Option[String],
-  access: Access
-)
+    name: FieldName,
+    clazz: DescriptorType,
+    generics: Option[String],
+    access: Access,
+    internalRefs: Set[FullyQualifiedName]
+) extends RawSymbol {
+  override def fqn: String = name.fqnString
+}
 
 final case class RawMethod(
-  name: MethodName,
-  access: Access,
-  generics: Option[String],
-  line: Option[Int]
-)
+    name: MethodName,
+    access: Access,
+    generics: Option[String],
+    line: Option[Int],
+    internalRefs: Set[FullyQualifiedName]
+) extends RawSymbol {
+  override def fqn: String = name.fqnString
+}
+
+sealed trait RawScalapSymbol {
+  def declaredAs: DeclaredAs
+  def access: Access
+  def scalaName: String
+  def typeSignature: String
+}
 
-final case class RawScalaClass(
+final case class RawScalapClass(
   javaName: ClassName,
   scalaName: String,
   typeSignature: String,
   access: Access,
   declaredAs: DeclaredAs,
-  fields: Seq[RawScalaField],
-  methods: Seq[RawScalaMethod]
-)
+  fields: Map[String, RawScalapField],
```

Comment:
using a `Seq[(String, RawScalapField)]` may not be glamorous, but it could also speed this up. I think I would prefer the `RawScalapField` to contain its name. Is the `String` here the bytecode name or the scala name? We need some convention/documentation on this or it will drive us crazy.